### PR TITLE
fix: get rid of stale nodes before going through disk

### PR DESCRIPTION
**Pull request by hadjiszs**
# Description

- the read smt db operation was not working for smt containing stale nodes
- fixes:
  - make the read smt db operation robust against smt with stale nodes
  - prune the stale nodes before each and every pass through the DB

## PR Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added or updated tests that comprehensively prove my change is effective or that my feature works

### CODE REVIEW

File: crates/agglayer-certificate-orchestrator/src/network_task.rs [hunk 1]
```diff
@@ -508,14 +513,21 @@ where
...
+                    if let Some(mut new) = self.pending_state.take() {
                         debug!(
                             "Updated the state for network {} with the new state {} > {}",
                             self.network_id,
                             self.local_state.get_roots().display_to_hex(),
                             new.get_roots().display_to_hex()
                         );
 
+                        // Prune the SMTs of the state
+                        new.prune_stale_nodes()
```
Comment (Freyskeyd) [inline crates/agglayer-certificate-orchestrator/src/network_task.rs:525]: Can we do that in the `cerifier` instead?

File: crates/agglayer-storage/src/stores/state/mod.rs [hunk 2]
```diff
@@ -396,9 +398,16 @@ impl StateStore {
                         left: *left.as_bytes(),
                         right: *right.as_bytes(),
                     });
-
-                    keys.push_back(SmtKeyType::Node(left));
-                    keys.push_back(SmtKeyType::Node(right));
+                    // update here
```
Comment (Freyskeyd) [inline crates/agglayer-storage/src/stores/state/mod.rs:401]:
```suggestion

```

File: crates/agglayer-storage/src/stores/state/mod.rs [hunk 3]
```diff
@@ -396,9 +398,16 @@ impl StateStore {
...
+                    if !already_read.contains(&left) {
+                        already_read.insert(left);
+                        keys.push_back(SmtKeyType::Node(left));
+                    }
+                    if !already_read.contains(&right) {
+                        already_read.insert(right);
+                        keys.push_back(SmtKeyType::Node(right));
+                    }
+                    // update end
```
Comment (Freyskeyd) [inline crates/agglayer-storage/src/stores/state/mod.rs:410]:
```suggestion
```

File: crates/agglayer-storage/src/stores/state/mod.rs [hunk 4]
```diff
@@ -381,6 +382,7 @@ impl StateStore {
         let mut nodes: Vec<Node<Keccak256Hasher>> = Vec::new();
         nodes.push(root_node);
 
+        let mut already_read = BTreeSet::new();
```
Comment (Freyskeyd) [inline crates/agglayer-storage/src/stores/state/mod.rs:385]: Maybe we can find a better name, not a blocker tho

File: crates/agglayer-storage/src/stores/state/mod.rs [hunk 5]
```diff
@@ -485,11 +494,21 @@ impl StateReader for StateStore {
         &self,
         network_id: NetworkId,
     ) -> Result<Option<LocalNetworkStateData>, Error> {
-        match (
-            self.read_local_exit_tree(network_id)?,
-            self.read_smt::<BalanceTreePerNetworkColumn, LOCAL_BALANCE_TREE_DEPTH>(network_id)?,
-            self.read_smt::<NullifierTreePerNetworkColumn, NULLIFIER_TREE_DEPTH>(network_id)?,
-        ) {
+        debug!("Reading local network state for network_id: {}", network_id);
```
Comment (Freyskeyd) [inline crates/agglayer-storage/src/stores/state/mod.rs:497]: I don't think we need to keep all of those debug right?

File: crates/agglayer-storage/src/stores/state/tests.rs [hunk 6]
```diff
@@ -135,3 +137,95 @@ fn can_detect_inconsistent_state(network_id: NetworkId, store: StateStore) {
...
+
+use pessimistic_proof_test_suite::sample_data::{self as data};
+
+#[rstest]
+fn can_read(network_id: NetworkId, store: StateStore) {
+    let certificates: Vec<Certificate> = [
+        "n15-cert_h0.json",
+        "n15-cert_h1.json",
+        "n15-cert_h2-v2.json",
+        //"n15-cert_h3-v2.json",
```
Comment (Freyskeyd) [inline crates/agglayer-storage/src/stores/state/tests.rs:149]: Should it be uncommented or is it a leftover?

**Review by Freyskeyd — APPROVED** [review]

File: crates/agglayer-storage/src/stores/state/mod.rs (hunk 4 above)
Comment (hadjiszs) [reply crates/agglayer-storage/src/stores/state/mod.rs]: I renamed it to `queued`, happy to rename it to something else

File: crates/agglayer-storage/src/stores/state/mod.rs (hunk 5 above)
Comment (hadjiszs) [reply crates/agglayer-storage/src/stores/state/mod.rs:493]: Removed, I kept only the 1st log

File: crates/agglayer-storage/src/stores/state/tests.rs (hunk 6 above)
Comment (hadjiszs) [reply crates/agglayer-storage/src/stores/state/tests.rs]: Leftover that I removed, this certificate was an attempt to replicate the h3 issue, I also renamed `n15-cert_h2-v2.json` to `n15-cert_h2.json` and deleted the unused ones

File: crates/agglayer-storage/src/stores/state/mod.rs (hunk 4 above)
Comment (Freyskeyd) [reply crates/agglayer-storage/src/stores/state/mod.rs]: I think that `queue` isn't the right name either, we're storing the `hash` of every node that the SMT contains, shouldn't it be something like `smt_hashes` or `node_hashes` ?
I also think that we could do the `if insert..` before even being in the match, but I'm fine letting it this way for now.

**Pull request merged**